src/note.rs

```diff
@@ -0,0 +1,394 @@
+                    &shared_secret,
+                    &nonce,
+                );
+
+                // TODO: replace the internals of PoseidonCipher with a plain
+                // array.
+                // Adding an infallible `to_bytes` method too.
+                let mut encrypted_data = [0u8; 96];
+                cipher
+                    .read_exact(&mut encrypted_data)
+                    .expect("Cannot encode PoseidonCipher");
+                encrypted_data
+            }
+        };
+
+        Note {
+            note_type,
+            value_commitment,
+            nonce,
+            stealth_address,
+            pos,
+            encrypted_data,
+        }
+    }
+
+    fn decrypt_data(&self, vk: &ViewKey) -> (u64, JubJubScalar) {
+        let R = self.stealth_address.R();
+        let shared_secret = dhke(vk.a(), R);
+        let nonce = BlsScalar::from(self.nonce);
+
+        // TODO: replace the internals of PoseidonCipher with a plain
+        // array.
+        let mut cipher = PoseidonCipher::default();
+
+        // TODO: improve the error handling
+        cipher
+            .write(&self.encrypted_data[..])
+            .expect("Cannot write the PoseidonCipher");
+
+        let data = cipher.decrypt(&shared_secret, &nonce);
+
+        // TODO: need to figure out what we want to do in case the decryption
+        // fails
+        return if data.is_err() {
+            (0u64, JubJubScalar::zero())
+        } else {
+            let data = data.unwrap();
+            let bytes = data[0].to_bytes();
+            let value = u64::from_le_bytes(chunk_of!(8; bytes[..8]));
+            let blinding_factor =
+                JubJubScalar::from_bytes(&data[1].to_bytes()).unwrap();
+            (value, blinding_factor)
+        };
+    }
+
+    /// Create a unique nullifier for the note
+    pub fn gen_nullifier(&self, sk: &SecretSpendKey) -> BlsScalar {
+        let sk_r = sk.sk_r(&self.stealth_address);
+        let sk_r = BlsScalar::from(sk_r);
+        let pos = BlsScalar::from(self.pos());
+
+        sponge_hash(&[sk_r, pos])
+    }
+
+    /// Return a hash represented by `H(value_commitment, pos, H([R]),
+    /// H([pskr]))`
+    pub fn hash(&self) -> BlsScalar {
+        let pk_r = JubJubAffine::from(self.stealth_address.pk_r());
+
+        sponge_hash(&[
+            self.value_commitment().get_x(),
+            self.value_commitment().get_y(),
+            BlsScalar::from(self.pos()),
+            pk_r.get_x(),
+            pk_r.get_y(),
+        ])
+    }
+
+    /// Return the type of the note
+    pub fn note(&self) -> NoteType {
+        self.note_type
+    }
+
+    /// Return the position of the note on the tree.
+    pub fn pos(&self) -> u64 {
+        self.pos
+    }
+
+    /// Set the position of the note on the tree.
+    /// This, naturally, won't reflect immediatelly on the data storage
+    pub fn set_pos(&mut self, pos: u64) {
+        self.pos = pos;
+    }
```

Comment:
Will this get the Index returned by the storage call from kelvin and store it?
If that's the case, this looks correct to me. Otherways doesn't since you can't decide where kelvin stores.

Reply:
Basically yes. This was used by the function `store_unspent_note` in the original code, and was set by the DB (Kelvin).

It's not used yet because that part is still missing (either `Rusk` that will take care of it or `phoenix-core` once is clear how the PoseidonTree are working from `Rusk-VM`).